importers/rules.go

```diff
@@ -0,0 +1,256 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package importers
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"math"
+	"net/url"
+	"os"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/api"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	plabels "github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/rulefmt"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/prometheus/prometheus/rules"
+	"github.com/prometheus/prometheus/tsdb/importer/blocks"
+)
+
+// RuleImporter is the importer for rules
+type RuleImporter struct {
+	logger log.Logger
+
+	config RuleConfig
+	groups map[string]*rules.Group
+
+	apiClient v1.API
+
+	writer *blocks.MultiWriter
+}
+
+// RuleConfig is the config for the rule importer
+type RuleConfig struct {
+	Start        string
+	End          string
+	EvalInterval time.Duration
+	URL          string
+}
+
+// NewRuleImporter creates a new rule importer
+func NewRuleImporter(logger log.Logger, config RuleConfig) *RuleImporter {
+	return &RuleImporter{
+		config: config,
+	}
+}
+
+// Init initializes the rule importer which creates a new block writer
+// and creates an Prometheus API client
+func (importer *RuleImporter) Init() error {
+	// create new block writer
+	newBlockDir, err := ioutil.TempDir("", "rule_blocks")
+	if err != nil {
+		return err
+	}
+	importer.writer = blocks.NewMultiWriter(importer.logger, newBlockDir, importer.config.EvalInterval.Nanoseconds())
+
+	// create api client
+	config := api.Config{
+		Address: importer.config.URL,
+	}
+	c, err := api.NewClient(config)
+	if err != nil {
+		return err
+	}
+	importer.apiClient = v1.NewAPI(c)
+	return nil
+}
+
+// Close cleans up any open resources
+func (importer *RuleImporter) Close() error {
+	return importer.writer.Close()
+}
+
+// Parse parses the groups and rules from a list of rules files
+func (importer *RuleImporter) Parse(ctx context.Context, files []string) (errs []error) {
+	groups := make(map[string]*rules.Group)
+
+	for _, file := range files {
```

Comment:
We seem to be duplicating code with the manager here

Reply:
i'm trying to de-duplicate the code and generalize the manager code so I can use it here as well, but its not going well... getting complicated. If you have any suggestions how to do it, let me know. The only thing I can come up with is that the rules.Manager might need a refactor to separate these different responsibilities.

Reply:
on second thought, I wonder if I can get away without needing a lot of this logic. It seems like all I really need to know in order to create new time series for recording rules is the query expression, group interval, and the info to create the alignment. Let me try to pair this down to only the data needed.